waveorder/io/writer.py

```diff
@@ -0,0 +1,297 @@
+        pass
+
+    def set_mm_metadata(self, meta):
+        # self.__builder.init_meta()
+        pass
+
+    def set_save_dir(self, path):
+        self._check_is_dir(path)
+
+    def set_store(self, path):
+
+        # Change to an existing store,
+        # if store doesn't exist, raise error
+
+        if os.path.exists(path):
+            print(f'Opening existing store at {path}')
+            self.store_path = path
+            self.store = zarr.open(path)
+        else:
+            raise ValueError(f'No store found at {path}, check spelling or create new store with add_store')
+
+    def init_array(self, data_shape, chunk_size, dtype):
+        self.__builder.init_array(self.store, data_shape, chunk_size, dtype)
+
+    # def set_stokes(self, stokes_data: np.ndarray):
+    #     # assign class attribute data
+    #     # call self.__builder attributes
+    #     self.__builder.init_arrays()
+    #     pass
+    #
+    # def set_physical(self, physical_data: np.ndarray):
+    #     # assign class attribute data
+    #     # desired keyword args:
+    #     #   data, channel, chunks, compressor
+    #     # can be learned from supplied keywords:
+    #     #   data.shape, chunk compatibility,
+    #     self.__builder.init_arrays()
+
+    # def write_stokes(self):
+    #     pass
+    #
+    # def write_physical(self):
+    #     pass
+
+    def write(self, data, P, T, Z):
+        """
+        Wrapper that calls the builder's write function.
+        Will write to existing array of zeros and place
+        data over the specified indicies
+
+        :param data: (nd-array), data to be saved. Must be the shape that matches indices (P, T, Z, Y, X)
+        :param P: (tuple or value), index or index range of the position dimension
+        :param T: (tuple or value), index or index range of the time dimension
+        :param Z: (tuple or value), index or index range of the Z dimension
+
+        """
+
+        self.__builder.init_zarr(self.store)
+        self.__builder.write(data, P, T, Z)
+        pass
+
+
+class Builder:
+    # interface for all builders
+
+    # create zarr memory store
+    def init_zarr(self): pass
+
+    # create subarrays named after the channel
+    def init_arrays(self): pass
+
+    # assign group and array attributes
+    def init_meta(self): pass
+
+    # create a directory store and write the memory store to that
+    def to_disk(self): pass
+
+
+class PhysicalZarr(Builder):
+    __pzarr = None
+    __compressor = None
+
+    def write(self, data, P, T, Z):
+        """
+        Write data to specified index of initialized zarr array
+
+        :param data: (nd-array), data to be saved. Must be the shape that matches indices (P, T, Z, Y, X)
+        :param P: (tuple or value), index or index range of the position dimension
+        :param T: (tuple or value), index or index range of the time dimension
+        :param Z: (tuple or value), index or index range of the Z dimension
```

Comment:
your use case earlier:
```
writer.write(data[0, 0, 0:65, :, :], P=[0], T=[0], Z=[0,65])
```
uses `lists` for P, T, Z.  So you will want to include that type in this docstring.

Additionally, calling `len(P)` when P is a "value" (assuming int) throws TypeError -- you'll need an explicit check for this.

Reply:
Agreed, check now implemented.